migrator/migrator.go

```diff
@@ -523,13 +522,105 @@ func (m Migrator) MigrateColumn(value interface{}, field *schema.Field, columnTy
 		}
 	}
 
-	if alterColumn && !field.IgnoreMigration {
-		return m.DB.Migrator().AlterColumn(value, field.DBName)
+	if alterColumn {
+		if err := m.DB.Migrator().AlterColumn(value, field.DBName); err != nil {
+			return err
+		}
+	}
+
+	if err := m.MigrateColumnUnique(value, field, columnType); err != nil {
+		return err
 	}
 
 	return nil
 }
 
+func (m Migrator) MigrateColumnUnique(value interface{}, field *schema.Field, columnType gorm.ColumnType) error {
+	unique, ok := columnType.Unique()
+	if !ok || field.PrimaryKey {
+		return nil // skip primary key
+	}
+
+	queryTx := m.DB.Session(&gorm.Session{})
+	execTx := queryTx
+	if m.DB.DryRun {
+		queryTx.DryRun = false
+		execTx = m.DB.Session(&gorm.Session{Logger: &printSQLLogger{Interface: m.DB.Logger}})
+	}
+
+	return m.RunWithValue(value, func(stmt *gorm.Statement) error {
+		// We're currently only receiving boolean values on `Unique` tag,
+		// so the UniqueConstraint name is fixed
+		constraint := m.DB.NamingStrategy.UniqueName(stmt.Table, field.DBName)
+		if m.UniqueAffectedByUniqueIndex {
+			if unique {
+				// Clean up redundant unique indexes
+				indexes, _ := queryTx.Migrator().GetIndexes(value)
+				for _, index := range indexes {
+					if uni, ok := index.Unique(); !ok || !uni {
+						continue
+					}
+					if columns := index.Columns(); len(columns) != 1 || columns[0] != field.DBName {
+						continue
+					}
+					if name := index.Name(); name == constraint || name == field.UniqueIndex {
+						continue
+					}
+					if err := execTx.Migrator().DropIndex(value, index.Name()); err != nil {
+						return err
+					}
+				}
+
+				hasConstraint := queryTx.Migrator().HasConstraint(value, constraint)
+				switch {
+				case field.Unique && !hasConstraint:
+					if field.Unique {
+						if err := execTx.Migrator().CreateConstraint(value, constraint); err != nil {
+							return err
+						}
+					}
+				// field isn't Unique but ColumnType's Unique is reported by UniqueConstraint.
+				case !field.Unique && hasConstraint:
+					if err := execTx.Migrator().DropConstraint(value, constraint); err != nil {
+						return err
+					}
+					if field.UniqueIndex != "" {
+						if err := execTx.Migrator().CreateIndex(value, field.UniqueIndex); err != nil {
+							return err
+						}
+					}
+				}
+
+				if field.UniqueIndex != "" && !queryTx.Migrator().HasIndex(value, field.UniqueIndex) {
+					if err := execTx.Migrator().CreateIndex(value, field.UniqueIndex); err != nil {
+						return err
+					}
+				}
+			} else {
+				if field.Unique {
+					if err := execTx.Migrator().CreateConstraint(value, constraint); err != nil {
+						return err
+					}
+				}
+				if field.UniqueIndex != "" {
+					if err := execTx.Migrator().CreateIndex(value, field.UniqueIndex); err != nil {
+						return err
+					}
+				}
+				return nil
+			}
+		} else {
+			if unique && !field.Unique {
```

Comment:
Do we need to handle this situation?
```diff
-	Name  string `gorm:"type:varchar(20);uniqueIndex:udx_name_name1"`
-	Name1 string `gorm:"type:varchar(20);uniqueIndex:udx_name_name1"`
+	Name  string `gorm:"type:varchar(20);uniqueIndex"`
+	Name1 string `gorm:"type:varchar(20);unique"`
```

Reply:
> What do you think is better?

When the unique of the field is equal, it seems that the index needs to be checked. If the Columns of the index do not contain the name of the field, the unique index may need to be deleted? I haven't researched it yet, because it may be more complicated, such as how to deal with adding or deleting a field in the composite index.
Of course, this doesn't have to be done in a PR, it's just a question, and I think it needs to be investigated and concluded.

Reply:
> When the unique of the field is equal, it seems that the index needs to be checked. 

It's not,  because `UniqueAffectedByUniqueIndex` is false.
In here, even if there is a UniqueIndex, it reports that Unique in ColumnType is false.

When `UniqueAffectedByUniqueIndex` is false:

```diff
- Name  string `gorm:"type:varchar(20);uniqueIndex:udx_name_name1"`
+ Name  string `gorm:"type:varchar(20);uniqueIndex"`
```
only the UniqueIndex names are different (not handle yet)

```diff
- Name1 string `gorm:"type:varchar(20);uniqueIndex:udx_name_name1"`
+ Name1 string `gorm:"type:varchar(20);unique"`
```
only add Unique constraints.


Behaviors: (That's actually what the test is about)

|  From | To |  UniqueAffected<br/>ByUniqueIndex | Behavior in MigrateColumnUnique |
| -- | -- | -- | -- |
| `gorm:"size:10"` | `gorm:"size:30;unique"` | true/false | add unique constraint |
| `gorm:"size:10"` | `gorm:"size:50;uniqueIndex"` | true | add unique index |
| `gorm:"size:10"` | `gorm:"size:50;uniqueIndex"` | false | - |
| `gorm:"size:30;unique"` | `gorm:"size:10"` | true/false | drop unique constraint |
| `gorm:"size:30;unique"` | `gorm:"size:50;uniqueIndex"` | true | drop unique constraint<br/>add unique index |
| `gorm:"size:30;unique"` | `gorm:"size:50;uniqueIndex"` | false | drop unique constraint |
| `gorm:"size:50;uniqueIndex"` | `gorm:"size:10"` | true | drop unique index |
| `gorm:"size:50;uniqueIndex"` | `gorm:"size:10"` | false | - |

Reply:
Once `UniqueAffectedByUniqueIndex`  is false, we don't care about UniqueIndex